server/src/main/java/org/opensearch/indices/replication/SegmentReplicationState.java

```diff
@@ -119,57 +225,303 @@ protected void validateAndSetStage(Stage expected, Stage next) {
+
+        if (targetNode != null) {
+            builder.startObject(Fields.TARGET);
+            builder.field(Fields.ID, targetNode.getId());
+            builder.field(Fields.HOST, targetNode.getHostName());
+            builder.field(Fields.TRANSPORT_ADDRESS, targetNode.getAddress().toString());
+            builder.field(Fields.IP, targetNode.getHostAddress());
+            builder.field(Fields.NAME, targetNode.getName());
+            builder.endObject();
+        }
+        builder.startObject(Fields.REPLICATING_STAGE);
+        replicating.toXContent(builder, params);
+        builder.endObject();
+
+        builder.startObject(Fields.GET_CHECKPOINT_INFO_STAGE);
+        getCheckpointInfo.toXContent(builder, params);
+        builder.endObject();
+
+        builder.startObject(Fields.FILE_DIFF_STAGE);
+        fileDiff.toXContent(builder, params);
+        builder.endObject();
+
+        builder.startObject(Fields.GET_FILES_STAGE);
+        index.toXContent(builder, params);
+        getFile.toXContent(builder, params);
+        builder.endObject();
+
+        builder.startObject(Fields.FINALIZE_REPLICATION_STAGE);
+        finalizeReplication.toXContent(builder, params);
+        builder.endObject();
+
+        return builder;
+    }
+
+    static final class Fields {
+        static final String ID = "id";
+        static final String REPLICATION_ID = "replication_id";
+        static final String STAGE = "stage";
+        static final String START_TIME = "start_time";
+        static final String START_TIME_IN_MILLIS = "start_time_in_millis";
+        static final String STOP_TIME = "stop_time";
+        static final String STOP_TIME_IN_MILLIS = "stop_time_in_millis";
+        static final String TOTAL_TIME = "total_time";
+        static final String TOTAL_TIME_IN_MILLIS = "total_time_in_millis";
+        static final String SOURCE = "source";
+        static final String HOST = "host";
+        static final String TRANSPORT_ADDRESS = "transport_address";
+        static final String IP = "ip";
+        static final String NAME = "name";
+        static final String TARGET = "target";
+        static final String INIT_STAGE = "init_stage";
+
+        static final String SIZE_OF_FILES_IN_BYTES = "size_of_files_in_bytes";
+
+        static final String INDEX = "index";
+        static final String TOTAL_GET_FILES_STAGE_TIME_IN_MILLIS = "total_get_files_stage_in_millis";
+        static final String REPLICATING_STAGE = "replicating_stage";
+        static final String GET_CHECKPOINT_INFO_STAGE = "get_checkpoint_info_stage";
+        static final String FILE_DIFF_STAGE = "file_diff_stage";
+        static final String GET_FILES_STAGE = "get_files_stage";
+        static final String FINALIZE_REPLICATION_STAGE = "finalize_replication_stage";
+    }
+
+    /**
+     * Starts Replicating
+     *
+     * @opensearch.internal
+     */
+    public static class Replicating extends ReplicationTimer implements ToXContentFragment, Writeable {
```

Comment:
Yes for now `List<Tuple<String, Long>> timingData;` will be enough to populate toXContent but in future we might capture more than just timing information from each stage of segment replication event. Keeping that in mind and using references of `RecoveryState class` I created all these new classes.

I don't have any strong opinions for either of these way. But if we have classes it would be easier to add and collect more information metrics in future.